big 'ol www update

rvagg opened the pull request:
For new nodejs.org server, 16.04 based rather than 14.04

* upstart -> systemd
* extracted https://github.com/nodejs/nodejs-dist-indexer and https://github.com/nodejs/nodejs-nightly-builder
* proper ssh keys
* remove cloudfuse, unused
* put benchmarking nginx, directory and ssh config in
* replace a lot of manual secret replacements with .j2 templates for much wow
* node v6 for running scripts (dist-indexer, nightly-builder and others)
* nightly-builder in crontab, although the branches are hardwired atm

rvagg commented:
I also have this in ~root for synchronizing data from the old server:

```
#!/bin/sh

rsync -avz --times --perms root@104.131.173.199:/home/dist/ /home/dist/ --exclude .ssh
rsync -avz --times --perms root@104.131.173.199:/home/libuv/ /home/libuv/

#NOTE: this misses out on *access.log, they should be preserved on the old server until DNS gives up
rsync -avz --times --perms root@104.131.173.199:/var/log/nginx/ /var/log/nginx/ --exclude access.log --exclude error.log --exclude iojs.org-access.log --exclude iojs.org-error.log --exclude nodejs.org-access.log --exclude nodejs.org-error.log --exclude libuv.org-access.log --exclude libuv.org-error.log
```

Will probably need to do that last one some time after conversion to pick up straggler DNS holdouts (thankfully cloudflare will handle most of that for us!). It's important we preserve as much of the log data for metrics even while we're serving from two hosts.

rvagg commented:
@nodejs/website @nodejs/benchmarking @saghul

We've set up a new server to host our various sites and have currently turned it on for benchmarking.nodejs.org and iojs.org. We'd also like dist.libuv.org updated to 138.197.224.240, @saghul are you the one to do that? AAAA is 2604:a880:400:d0::b2c:a001 if you have one of those.

We'll probably switch over nodejs.org to this new host (fronted by CloudFlare still) tomorrow, for now we're easing in. Please watch for problems, bug reports, people complaining, the usual signs of something wrong and let us know. Ping myself or @jbergstroem directly, IRC is a good place to get us both in a timely manner.

Will update here when we move nodejs.org over.

rvagg commented:
@jbergstroem I think I'm pretty much done with this PR, also added:

* swap spdy for http2
* added metrics setup
* extracted latest-linker to https://github.com/nodejs/nodejs-latest-linker

should be simpler to translate it all to your new work without all the additional config. resources/secrets/makesecrets.sh should make the secrets for you although there's stuff in host_vars that doesn't get made by that script, not sure how to deal with the extra stuff, I guess somewhere in the secrets repo?

jbergstroem commented on setup/www/tasks/site-setup.yaml:
```diff
@@ -10,39 +10,55 @@
+  tags: setup
+
 - name: Site Setup | Copy cdn-purge.sh script
   copy:
```
this needs to be template if you're using variables.

jbergstroem commented on setup/www/tasks/site-setup.yaml:
```diff
@@ -81,3 +84,16 @@
+    - '*  *    * * *   nodejs  /home/nodejs/check-build-site.sh nodejs'
+    - '*  *    * * *   nodejs  /home/nodejs/check-build-site.sh iojs'
+    - '*  *    * * *   root    /home/nodejs/cdn-purge.sh'
+    - '0 */4   * * *   nodejs  rsync -avz --delete benchmark:charts/ /home/www/benchmarking/charts/'
```
perhaps add a nightly cronjob that updates the npm packages? similar to https://github.com/jbergstroem/build/blob/feature/refactor-the-world/ansible/playbooks/jenkins/linter.yml#L36..L40.

silverwind commented on setup/www/resources/config/benchmarking.nodejs.org:
```diff
@@ -0,0 +1,38 @@
+    resolver 8.8.4.4 8.8.8.8 valid=300s;
+    resolver_timeout 10s;
+
+    add_header Strict-Transport-Security max-age=63072000;
```
Could add `preload` here and submit the site to https://hstspreload.appspot.com/ once it goes live.

jbergstroem replied:
also: this needs to be commented until we've landed #484.

silverwind replied:
Ah, looks like a complex situation there, feel free delay preload enablement at a later point, as it is certainly a commitment that cannot be undone.

rvagg replied:
might be a bit too complex for this, we have subdomains that we want to still serve on port 80, including unencrypted.nodejs.org

silverwind commented on setup/www/resources/config/benchmarking.nodejs.org:
```diff
@@ -0,0 +1,38 @@
+    gzip on;
+    gzip_static on;
+    gzip_disable "MSIE [1-6]\.";
+    gzip_types text/plain text/css application/x-javascript text/xml application/xml application/xml+rss text/javascript;
```
WOFF fonts are already highly compressed, so no sense in gzipping them. `image/svg+xml` should be added, yes. Oh, and I don't think anyone would mind if you drop the IE6 hack above.

silverwind replied:
Oh, and I think the two javascript mimes can be reduced to just `application/javascript`.

rvagg commented:
* added svg and changed to `application/javascript` in `gzip_types`
* added cron job to `npm update -g` @jbergstroem pls have a look
* combined the 3 JS tools into a single tools.yaml

jbergstroem commented:
LGTM

rvagg commented:
Task list for transition (shortly):

- [ ] CloudFlare DNS update nodejs.org, direct.nodejs.org to 138.197.224.240 and AAAA to 2604:a880:400:d0::b2c:a001 (tho I don't know if this really does anything via CF)
- [ ] Update GitHub webhook endpoint for nodejs.org and iojs.org to the new server (if they have a hardwired IP then I'll change that to direct.nodejs.org)
- [ ] Run a test push to nodejs.org master to make sure it updates automatically
- [ ] Run a nightly build to check if CI release servers can access the new host
- [ ] Run a release build and have a releaser (who is not me) attempt to promote and check that they can at least get a list of builds available
- [ ] Point backup server to the new host
- [ ] Change rsnapshot config to fetch log files from subdirectories of /var/log/nginx (have moved each hostname to a subdir, we can exclude old/ but I'd still like those kept somewhere other than this server even if they are a static once-off)
- [ ] Copy remaining logs off old server to new server as DNS drains (give it ~48 hours at least although most traffic is through CF so it'll be pretty quick)

@jbergstroem pls update my comment with any additional items you can think of

rvagg commented:
@nodejs/collaborators ping, we're just about to switch over nodejs.org to a new backend server, please stay tuned to chatter about problems with the site and let us know ASAP if there are any concerns. We don't expect any of course but you never know!

saghul commented:
Thanks for the heads up, I emails @piscisaureus since he handles the libuv DNS, hopefully that is adjusted shortly.

piscisaureus commented:
done

On Wed, Oct 12, 2016 at 10:54 PM, Saúl Ibarra Corretgé <
notifications@github.com> wrote:

> Thanks for the heads up, I emails @piscisaureus
> <https://github.com/piscisaureus> since he handles the libuv DNS,
> hopefully that is adjusted shortly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/build/pull/516#issuecomment-253420866>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AANUkSEMsFk2-NG8rGBS2AYVD-nd_X8qks5qzceBgaJpZM4KUSIv>
> .
>

rvagg commented:
Just discovered the hard way that the promotion scripts were out of date and didn't handle `test` builds or even 0.10 and 0.12. I must have a pull request somewhere with an update that never made it in to master. Updated in latest commit.

rvagg commented:
Added coverage.nodejs.org as per https://github.com/nodejs/testing/issues/36

jbergstroem commented:
Since you're in here, can you add the unencrypted pubkey to the root user? I guess we should use iojs or something but that's at least how it works now.

cjihrig commented:
I am getting reports on IRC that the libuv downloads are only available over IPv4, and returning a 301 redirect over IPv6. I'm not sure if this is the right place to bring it up, or if this issue is even related. I'm just curious if anyone from the build working group can help shed some light. Thanks.

jbergstroem commented:
@cjihrig by looking at the config attached to this PR I would say that sounds about right: https://raw.githubusercontent.com/nodejs/build/a49c891f3b4180631a174e28056bd0395232dec8/setup/www/resources/config/libuv.org

@rvagg perhaps:
```diff
--- libuv.org   2016-10-25 12:00:04.000000000 -0300
+++ libuv.org-ipv6      2016-10-25 12:01:09.000000000 -0300
@@ -1,5 +1,6 @@
 server {
     listen *:80;
+    listen [::]:80;
     server_name dist.libuv.org;
 
     keepalive_timeout 60;
```

rvagg commented:
k, have done that on http://dist.libuv.org/, http://coverage.nodejs.org/ (just the port 80 redirect) and http://benchmarking.nodejs.org/ (ditto port 80 redirect), if someone with a decent ipv6 setup could report back on how well that's working it'd be appreciated.

txdv commented:
txdv | Dani-hp: can you test now if it works correctly?
    Dani-hp | yep looking good now
    Dani-hp | thanks guys

The guy reporting the problem says it is working now.

AMDmi3 commented:
I also confirm the fix :+1:
